instrumentctl/g9_driver.py

```diff
@@ -0,0 +1,264 @@
+import serial
+import threading
+import time
+from utils import LogLevel
+import os
+from subsystem import interlocks
+
+inStatus = {
+    0: "No error",
+    1: "Invalid configuration",
+    2: 'External test signal failure',
+    3: 'Internal circuit error',
+    4: 'Discrepancy error',
+    5: 'Failure of the associated dual-channel input'
+}
+
+outStatus = {
+    0: 'No error',
+    1: 'Invalid configuration',
+    2: 'Overcurrent detection',
+    3: 'Short circuit detection',
+    4: 'Stuck-at-high detection',
+    5: 'Failure of the associated dual-channel output',
+    6: 'Internal circuit error',
+    8: 'Dual channel violation'
+}
+
+usStatus = {
+    9: "Output Power Supply Error Flag",
+    10: "Safety I/O Terminal Error Flag",
+    13: "Function Block Error Flag"
+}
+
+class G9Driver:
+    def __init__(self, port=None, baudrate=9600, timeout=0.5, logger=None, debug_mode=False):
+        if port:
+            self.ser = serial.Serial(port, baudrate, parity=serial.PARITY_EVEN, stopbits=serial.STOPBITS_ONE, bytesize=serial.EIGHTBITS, timeout=timeout)
+        self.debug_mode = debug_mode
+        self.logger = logger
+        self.lastResponse = None
+        self.msgOptData = None
+        self.input_flags = None
+
+    def send_command(self):
+        if not self.is_connected():
+            raise ConnectionError("Seiral Port is Not Open.")
+        header = b'\x40\x00\x00\x0F\x4B\x03\x4D\x00\x01' 
+        data = b'\x00\x00\x00\x00'
+        reserve = b'\x00\x00'
+        self.msgOptData = data
+        checksum = self.calculate_checksum(header + data + reserve, 0 , len(header + data))
+        cs = b'\x00' + checksum if len(checksum) == 1 else checksum
+        footer = b'\x2A\x0D' 
+        #TODO: add try and catch
+        self.ser.write(header + data + reserve+ cs + footer)
+
+        self.response()
+
+    # used mainly for the check sum but can also be used to check for error flags
+    # needs an input of a byte string and the range of bytes that need to be sum
+    # will return the sum of the bytes in the a byte string in the form of b'\x12'
+    def calculate_checksum(self, byteString, startByte, endByte):
+        assert isinstance(byteString, bytes)
+        return sum(byteString[startByte:endByte + 1]).to_bytes(1, "big") 
```

Comment:
guys I think I see an issue with this checksum method. Has anyone confirmed if this actually works as expected?

Pg. 116 of the datasheet specifies that the checksum is 16 bits (2 bytes) long. The `to_bytes` built-in in this line explicitly sets the length to **1 byte**:
```python
return sum(byteString[startByte:endByte + 1]).to_bytes(1, "big")
```

I think this would truncate the checksum to the least significant 8 bits, and discard any higher bits?

Reply:
> guys I think I see an issue with this checksum method. Has anyone confirmed if this actually works as expected?
> 
> Pg. 116 of the datasheet specifies that the checksum is 16 bits (2 bytes) long. The `to_bytes` built-in on this line explicitly sets the length to **1 byte**:
> 
> ```python
> return sum(byteString[startByte:endByte + 1]).to_bytes(1, "big")
> ```
> 
> I think this would truncate the checksum to the least significant 8 bits, and discard any higher bits?

They did get the checksum to work with the test code which was last week which I assume is either similar or the same. The G9 has 2 bytes for the checksum, but the checksum with the zero optional data we send is always fully contained within the low byte, which is likely why they throw away the high byte. This may be poor practice though, for example, if we decided to send optional data at some point I haven't done the math to determine if we could end up with non zero high byte data. I think we should also calculate the full check sum and send the high byte and the low byte, which in our case most of the time (possibly all) the high byte will be zero.

Reply:
okay yea that makes sense. In any case I think truncating the checksum undermines the purpose of the checksum. and we shouldn't rely on an assumption that the high byte will always be zero. esp because this is part of a safety system

Reply:
Yeah I wrote that, and tested it. I can do a little more follow up testing to make sure it works in all cases( including with optimal data). I will document these tests somewhere.